### DISCUSSION TITLE
Memory allocation guidelines when training splink on pyspark via EMR

### DISCUSSION THREAD

**Discussion by jkginfinite** [moj-analytical-services/splink#1355] [Q&A]
Hello,
My team is currently working with AWS EMR & pyspark with 6 nodes each with 11.7 gb of ram and 4 cpu-vcores

I am trying to perform de-duplication on a 30.3gb dataset pulled via either S3 parquet or MySQL. The dataset has about 5M rows with 33 columns of strings and other mixed data types. (personal identifiable information).

We have been successful at training and predicting on roughly 100k rows with 1gb driver memory and 1gb executor memory, but beyond this we get a "memory allocation" error.

I asked Chat GPT on the size amount of free memory on our cluster that should be available for a 30gb data set, and it said roughly 2-3x the size of the dataset.

**My question is: 

Is there a rule of thumb for the available memory needed for Splink to train a dataset - should it be, across all nodes like chat gpt says - 2-3x the size of the dataset?**

**Comment by RobinL** [comment]
Hi Joe.  



The closest I could get to a rule of thumb would be that I would expect you to be able generate something like 50 million record comparisons reasonably easily on your setup, and probably get quite a bit higher than that with careful attention to Spark and Splink settings.

Note that the fundamental constraint is usually the number of record comparisons you're generating using Splink as opposed to the size of the input dataset.

The best place in the docs to look are the following links:
[Run times, performance and linking large data](https://moj-analytical-services.github.io/splink/topic_guides/drivers_of_performance.html)
[Optimising Spark performance](https://moj-analytical-services.github.io/splink/topic_guides/optimising_spark.html)
and also more details on blocking [here](https://moj-analytical-services.github.io/splink/topic_guides/blocking_rules.html)

To summarise:
> your choice of blocking rules is by far the most important driver of performance.

I would probably start by running one of these two functions to work out how many comparisons are being generated:
https://moj-analytical-services.github.io/splink/linkerexp.html?h=count_num#splink.linker.Linker.count_num_comparisons_from_blocking_rule
or
https://moj-analytical-services.github.io/splink/linker.html?h=count#splink.linker.Linker.count_num_comparisons_from_blocking_rules_for_prediction

For a cluster of the size you're using, I'd start with blocking rules that generate about 10 million comparison pairs or fewer.  You should be able to go a lot higher than this (probably to about 100 million or so on your cluster).  But best to start lower and build up.

You also want to avoid significant skew (e.g. if you're blocking on first name and surname, you'll end up with far too many comparisons being generated within 'john smiths').  You can use [profile_columns](https://moj-analytical-services.github.io/splink/linker.html?h=profile_#splink.linker.Linker.profile_columns) to look for skew.

Finally, typically I think you'd set your executor memory considerably higher than 1gb on the size of machine you're using - to maybe 5gb.  I believe that's the default setting in AWS glue on 16gb machines, which we use.

**Reply by jkginfinite** [reply]
Thanks,

In order to run the functions;

I would probably start by running one of these two functions to work out how many comparisons are being generated:
https://moj-analytical-services.github.io/splink/linkerexp.html?h=count_num#splink.linker.Linker.count_num_comparisons_from_blocking_rule
or
https://moj-analytical-services.github.io/splink/linker.html?h=count#splink.linker.Linker.count_num_comparisons_from_blocking_rules_for_prediction

does the model need to be fully trained? 
Do blocking rules for predictions need to be specified in the settings before hand?

thanks for your help

**Reply by RobinL** [reply]
No - those functions don't require a trained model to run.  They analyse the input rows to compute how many comparisons would be generated by the final model.

For the first one, you provide the blocking rule as an argument, so it's independent from your settings.

For the second function, it looks at the settings you've provided, and uses the blocking rules for prediction that you're specified there.

One final note:  if your blocking rules are very loose (for example if you have a blocking rule that would generate tens of billions of comparison pairs), even the function to count them may take a long time to run.  Which is to say if those functions are taking a long time to run and not returning an answer, then it is a good indication you need stricter blocking rules.

**Reply by jkginfinite** [reply]
thanks. Just ran these functions in spark EMR. Was able to load run the code and specify settings for the linker.

Noticed an odd thing happened - I tried looping through a set of rules, and the code just above the linker.count_num_comparisons_from_blocking_rule(rule) ran, but that function with the rule, was skipped and so did the lines below it in the loop.

Any reason why the function linker.count_num_comparisons_from_blocking_rule(rule) could be skipped? 

I'm guessing its because EMR didn't fully load the dataset maybe?
Working with EMR. Its as line of code isn't even there.

**Comment by RobinL** [comment]
@JoeGanser been meaning to dig this out - this is a listing of the default spark config of executors when you use AWS Glue - i.e. these are the settings we (implicitly, because they're just the defaults) use in production.  Note we're using the [standard worker type](https://docs.aws.amazon.com/glue/latest/dg/add-job.html#:~:text=custom%20scripts.-,Worker%20type,-The%20following%20worker).

> Each worker maps to 1 DPU (4 vCPUs, 16 GB of memory) with 84GB disk (approximately 34GB free)

This is a little old now because I dug it out of Slack from 2018!  I also note my colleague at the time asked the sensible question:
> So what I don’t understand is why the spark.executor.memory is set to 5g. I think each executor is a DPU which is meant to have 4 CPUS (which is correct from the spark.executor.cores value) and 16GB of memory
to which I don't have a good answer!

I slightly edited anything that looked potentially sensitive (I don't think any of it was, but better safe then sorry)

```
[('spark.dynamicAllocation.minExecutors', '1'),
 ('spark.hadoop.mapred.output.committer.class',
  'org.apache.hadoop.mapred.DirectOutputCommitter'),
 ('spark.submit.deployMode', 'cluster'),
 ('spark.driver.port', 'xx'),
 ('spark.driver.extraLibraryPath',
  '/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native'),
 ('spark.yarn.dist.jars', 'file:/opt/amazon/superjar/glue-assembly.jar'),
 ('spark.executor.memory', '5g'),
 ('spark.app.id', 'application_xx'),
 ('spark.driver.memory', '5g'),
 ('spark.executor.cores', '4'),
 ('spark.org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter.param.PROXY_HOSTS',
  'ip-xxx.eu-west-1.compute.internal'),
 ('spark.yarn.dist.pyFiles',
  'file:/tmp/PyGlue.zip, 'queries.py'),
 ('spark.executor.extraClassPath',
  './*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*'),
 ('spark.yarn.maxAppAttempts', '1'),
 ('spark.sql.warehouse.dir', 'hdfs:///user/spark/warehouse'),
 ('spark.serializer.objectStreamReset', '100'), 
 ('spark.hadoop.fs.defaultFS',
  'hdfs://xx.eu-west-1.compute.internal:xx'),
 ('spark.yarn.app.id', 'application_xx'),
 ('spark.ui.filters',
  'org.apache.hadoop.yarn.server.webproxy.amfilter.AmIpFilter'),
 ('spark.driver.extraClassPath',
  './*:/usr/lib/hadoop-lzo/lib/*:/usr/share/aws/emr/emrfs/conf:/usr/share/aws/emr/emrfs/lib/*:/usr/share/aws/emr/emrfs/auxlib/*'),
 ('spark.shuffle.service.enabled', 'true'),
 ('spark.app.name', 'tape'),
 ('spark.hadoop.yarn.resourcemanager.address',
  'ip-xxx.eu-west-1.compute.internal:xx'),
 ('spark.executor.id', 'driver'),
 ('spark.hadoop.yarn.resourcemanager.connect.max-wait.ms', '60000'),
 ('spark.executorEnv.PYTHONPATH',
  '{{PWD}}/__pyfiles__<CPS>{{PWD}}/pyspark.zip<CPS>{{PWD}}/py4j-0.10.4-src.zip<CPS>{{PWD}}/PyGlue.zip'),
 ('spark.hadoop.mapreduce.fileoutputcommitter.marksuccessfuljobs', 'false'),
 ('spark.hadoop.parquet.enable.summary-metadata', 'false'),
 ('spark.master', 'yarn'),
 ('spark.ui.port', '0'),
 ('spark.rdd.compress', 'True'),
 ('spark.executor.extraLibraryPath',
  '/usr/lib/hadoop/lib/native:/usr/lib/hadoop-lzo/lib/native'),
 ('spark.yarn.secondary.jars', 'glue-assembly.jar'),
 ('spark.yarn.isPython', 'true'),
 ('spark.dynamicAllocation.enabled', 'true'),
 ('spark.driver.host', 'xxx')]
```
